hpx/runtime/components/client_base.hpp

```diff
@@ -124,143 +169,183 @@ namespace hpx { namespace components
     class client_base : public detail::make_stub<Stub>::type
     {
+    private:
+        template <typename T, typename Enable>
+        friend struct hpx::traits::detail::future_access_customization_point;
+
     protected:
         typedef typename detail::make_stub<Stub>::type stub_type;
-        typedef shared_future<naming::id_type> future_type;
+        typedef lcos::detail::future_data<id_type> shared_state_type;
+
+        typedef shared_future<id_type> future_type;
+
+        void unregister_held_object(error_code& ec = throws)
+        {
+            if (!registered_name_.empty())
+            {
+                std::string name = std::move(registered_name_);
+                agas::unregister_name_sync(name, ec);
+            }
+        }
+
+        client_base(boost::intrusive_ptr<shared_state_type> const& state)
+          : shared_state_(state)
+        {}
+
+        client_base(boost::intrusive_ptr<shared_state_type> && state)
+          : shared_state_(std::move(state))
+        {}
 
     public:
         typedef Stub stub_argument_type;
         typedef typename detail::make_stub<Stub>::server_component_type
             server_component_type;
 
+        typedef void is_client_tag;
+
         client_base()
-          : gid_()
+          : shared_state_()
         {}
 
-        explicit client_base(naming::id_type const& gid)
-          : gid_(lcos::make_ready_future(gid))
-        {}
-        explicit client_base(naming::id_type && gid)
-          : gid_(lcos::make_ready_future(std::move(gid)))
-        {}
+        explicit client_base(id_type const& id)
+          : shared_state_(new shared_state_type)
+        {
+            shared_state_->set_value(id);
+        }
+        explicit client_base(id_type && id)
+          : shared_state_(new shared_state_type)
+        {
+            shared_state_->set_value(std::move(id));
+        }
 
-        explicit client_base(future_type const& gid)
-          : gid_(gid)
+        explicit client_base(shared_future<id_type> const& f) HPX_NOEXCEPT
+          : shared_state_(
+                hpx::traits::future_access<future_type>::
+                    get_shared_state(f))
         {}
-        explicit client_base(future_type && gid)
-          : gid_(std::move(gid))
+        explicit client_base(shared_future<id_type> && f) HPX_NOEXCEPT
+          : shared_state_(
+                hpx::traits::future_access<future_type>::
+                    get_shared_state(std::move(f)))
         {}
-        explicit client_base(future<naming::id_type> && gid)
-          : gid_(gid.share())
+        explicit client_base(future<id_type> && f) HPX_NOEXCEPT
+          : shared_state_(hpx::traits::future_access<future_type>::
+                    get_shared_state(std::move(f)))
         {}
 
-        explicit client_base(client_base const& rhs)
-          : gid_(rhs.gid_)
+        client_base(client_base const& rhs) HPX_NOEXCEPT
+          : shared_state_(rhs.shared_state_)
```

Comment:
Currently the `client_base` instance which registered the id will unregister it in ist destructor. This is not a good solution and Needs separate discussions. But for now we shouldnät copy the name as it would be unregistered twice if we did.